### DISCUSSION TITLE
How could I debug nvim config easily?

### DISCUSSION THREAD

**Discussion by qlibp** [nix-community/kickstart-nix.nvim#11] [Q&A] [answered]
if I modify the github repo's lua configuration locally, the behavior won't change, and it's sure a good thing. But is there a way to quickly write & evaluate the nvim configuration by just modifying the config directly?

Should I re-build flake every time when I change sth to see what happens?

**Comment by mrcjkb** [comment]
Hey :wave: 

My workflow is usually

- Perform modifications
- `nix run /path/to/flake#nvim` to test the modifications

Although it is a slightly slower feedback loop, I find it beneficial, because if something breaks severely, I can still use my unmodified config to fix things.

If you're working on something where you need a faster feedback loop, you can copy the module you're working on to your `$XDG_CONFIG_HOME/$NVIM_APPNAME/...` directory and work on it there. Then copy it back when you're done.
Make sure to filter out the module you've copied over from the derivation, so that it doesn't conflict.

**Comment by qlibp** [comment]
> working on to your `$XDG_CONFIG_HOME/$NVIM_APPNAME/...` directory and work on it there

I try this one just now, but nvim seems to not load my script.

I place a `init.lua` in `~/.config/nvim`, and nvim don't load this script at all. Check with `:scriptnames`

**Comment by mrcjkb** [comment]
> I place a `init.lua` in `~/.config/nvim`, and nvim don't load this script at all. Check with `:scriptnames`

It won't source an `init.lua`, because the neovim derivation has a wrapper which calls `nvim -u /path/to/generated-init.lua`.
You can run `nvim -u /path/to/init.lua`, but then it'll ignore the generated init.lua.

An alternative might be to put your script in `~/.config/nvim/plugin`. But in this case, it'll be sourced after the generated init.lua.

**Comment by qlibp** [comment]
This practice should be documented in README for those like me who's not so farmiliar with nix

**Comment by qlibp** [comment]
One more question, if I newly add a nvim config file, maybe `nvim/plugin/lspconfig.lua`, and then run `nix run .#nvim`, will this file get loaded?

Concretely, when I try to introduce pylsp into the configuration, the above procedure don't work. I then add the line 
`require('lspconfig').pylsp.setup {}` into `init.lua`, then all works fine. Therefore, I suspect that the newly added files is not loaded.

Correct me if I'm wrong.

**Reply by juangiordana** [reply]
This thread might be of help: https://github.com/Misterio77/nix-starter-configs/issues/68#issuecomment-2032024538

 I figured a more comfortable workflow/transition path by not to having an active git repository. Once you are comfortable you can start commiting your configs.

**Reply by mrcjkb** [reply]
Good idea.

I've added a section to the readme (that includes your tip @juangiordana)

**Reply by shofel** [reply]
That's so helpful! I spent unreasonable lots of time iterating on lua part. Then I employed a plugin to execute lua in theh buffer (lluadev-nvim plugin by bfredl).

But the ability to restart without reload is priceless. I takes 40s to `nix run .`. 40s to check every small edit? That's too much frustration.

I think, for myself, I'll even make it vice versa:
- use symlinked directory by default. For easier everyday iteration. That is, `nvim` will mean `nvim-dev`
- run with wrapRc=true as alternative, when I want a known good commited config. The name of the app could be :thinking: `nvim-` or `nvim-fromNixStore`

**Comment by mrcjkb** [comment, marked as answer]
With nix flakes, only files that have been staged (`git add`) or committed will be added to the derivation.
